### PR TITLE
Upstream PLIST handling

### PR DESCRIPTION
**Pull request by matthewbauer**
###### Motivation for this change
Need to make sure this is okay with @lnl7 first, but this is the plist generator from nix-darwin. Hopefully there is no problem with upstreaming it to Nixpkgs.

@3noch has some improvements for it that are needed by reflex-platform. I want to work with him to get this included in Nixpkgs as soon as possible (https://github.com/LnL7/nix-darwin/pull/78 still needs 
some changes).

### CODE REVIEW

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build darwin.adv_cmds

**Comment by matthewbauer** [comment]
Anyone know of a package that uses xcbuild & works in darwin?

File: lib/generators.nix [hunk 1]
```diff
@@ -175,4 +175,52 @@ rec {
                          else "<λ:{${showFnas}}>"
     else abort "toPretty: should never happen (v = ${v})";
 
+  # PLIST handling
+
+  toPLIST = x: ''
```
Comment (Profpatsch) [inline lib/generators.nix]: `toX` functions should all take an attrset before the input value, which is an attribute set of options (possibly added in the future). So

```
toPlist = {}: x: ''
```

I think `plist` is normally written with small letters (going by the [wiki article](https://en.wikipedia.org/wiki/Property_list)), so I’d camel case.

The next step is thinking about all parts of the generator and moving everything that is not fixed by the generated format into an option. For example `mkINI` doesn’t specify how to format names in section headers (mainly what escaping is used, `mkSectionName`) and does not fix how key-value-pairs look, only the fact that they exist (the formatting of them is relayed to `mkKeyValueDefault`, another generator).

File: lib/generators.nix [hunk 2]
```diff
@@ -175,4 +175,52 @@ rec {
...
+     + "\n</plist>";
+
+  pprExpr = ind: x: with builtins;
+    if isNull x then "" else
+    if isBool x then pprBool ind x else
+    if isInt x then pprInt ind x else
+    if isString x then pprStr ind x else
+    if isList x then pprList ind x else
+    if isAttrs x then pprAttrs ind x else
+    throw "invalid plist type";
```
Comment (Profpatsch) [inline lib/generators.nix]: The error message should mirror the other pattern matching error messages in this module, namely:

```
    else abort "toPretty: should never happen (v = ${v})";
```

and 

```
    let err = t: v: abort
          ("generators.mkValueStringDefault: " +
           "${t} not supported: ${toPretty {} v}");
```

And now that I see them, they can you edit factor out the error function from the second and make all say:

```
"generators.${name}: this value is not supported: ${toPretty {} v}"
```

? :)

File: lib/generators.nix [hunk 3]
```diff
@@ -175,4 +175,52 @@ rec {
...
+
+  pprAttrs = ind: x: libStr.concatStringsSep "\n" [
+    (pprLiteral ind "<dict>")
+    (pprAttr ind x)
+    (pprLiteral ind "</dict>")
+  ];
+
+  attrFilter = name: value: name != "_module" && value != null;
+
+  pprAttr = ind: x: libStr.concatStringsSep "\n" (lib.flatten (lib.mapAttrsToList (name: value: lib.optional (attrFilter name value) [
```
Comment (Profpatsch) [inline lib/generators.nix]: Line length

File: lib/generators.nix [hunk 4]
```diff
@@ -175,4 +175,52 @@ rec {
...
+  # PLIST handling
+
+  toPLIST = x: ''
+    <?xml version="1.0" encoding="UTF-8"?>
+    <!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+    <plist version="1.0">
+  '' + pprExpr "" x
+     + "\n</plist>";
+
+  pprExpr = ind: x: with builtins;
```
Comment (Profpatsch) [inline lib/generators.nix]: The `pprX` definitions should go into a `let` definition at the start of `toPlist`, and the `ppr` prefixes can be removed.

**Review by Profpatsch — CHANGES_REQUESTED** [review]
That’s all I have on reading through the `generators.nix` code.

In general, the generator seems to have some duplication, so maybe it can be cleaned up a bit.

As for the indentation trick, that’s pretty smart, I once had the same problem and arrived at a different solution, that feels more combinator-like, but was a bit brittle: https://gitlab.techcultivation.org/sangha/sangha-deployment/blob/master/containers/helpers/write-script-argparse/build-script.nix#L97
(No criticism, just for the fun of the problem).

**Review by LnL7 — COMMENTED** [review]
Like I mentioned on the nix-darwin pr, I'm fine with moving these to nixpkgs. Didn't realise there are some places where it could be used already.

**Comment by matthewbauer** [comment]
Ok is all caps PLIST acceptable? We can do toPlist as well if that's preferred.

File: lib/generators.nix [hunk 5]
```diff
@@ -175,4 +175,51 @@ rec {
                          else "<λ:{${showFnas}}>"
     else abort "toPretty: should never happen (v = ${v})";
 
+  # PLIST handling
+  toPLIST = {}: v: let
+    pprExpr = ind: x: with builtins;
```
Comment (Profpatsch) [inline lib/generators.nix]: The `ppr` prefix is superfluous.

File: lib/generators.nix [hunk 6]
```diff
@@ -175,4 +175,51 @@ rec {
                          else "<λ:{${showFnas}}>"
     else abort "toPretty: should never happen (v = ${v})";
 
+  # PLIST handling
+  toPLIST = {}: v: let
```
Comment (Profpatsch) [inline lib/generators.nix]: I’m torn. When I search for it, everyone else doesn’t write `PLIST`, but `plist`. So I’d prefer camel case in this case.

**Review by Profpatsch — CHANGES_REQUESTED** [review]

**Review by Profpatsch — APPROVED** [review]

**Pull request merged**